### DISCUSSION TITLE
MPEmbeddings result differs from sentence_transformers

### DISCUSSION THREAD

**Discussion by AlexLamaison-UnlikelyAI** [JohnSnowLabs/spark-nlp#14494] [Q&A] [answered]
I'm trying to Sparkify some code that previously used sentence_transformers, but the vectors I get back using MPEmbeddings differ from the ones i get back using sentence_transformers.  Why might this be?

The sentence_transformers code:
```
from sentence_transformers import SentenceTransformer
model = SentenceTransformer("all-mpnet-base-v2")
model.encode("Description: interlobar artery. Known as: 'interlobar artery'.").tolist()
```

and the Spark-NLP code:
```
from sparknlp.base import *
from sparknlp.annotator import *
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("Spark NLP MPNet") \
    .getOrCreate()
document = DocumentAssembler() \
    .setInputCol("text") \
    .setOutputCol("document")
mpnet = MPNetEmbeddings.pretrained("all_mpnet_base_v2", "en") \
    .setInputCols(["document"]) \
    .setOutputCol("embeddings")

pipeline = Pipeline(stages=[document, mpnet])

data = spark.createDataFrame([["Description: interlobar artery. Known as: 'interlobar artery'."]], ["text"])
result = pipeline.fit(data).transform(data)
result.show(truncate=False)
```

I would expect these to produce the same values because they use the same model, but they don't.  What am I missing?

**Comment by AlexLamaison-UnlikelyAI** [comment]
From running bits of the two implementations in REPLs, it looks like Spark-NLPs MPNetEmbeddings tokenizer doesn't add tokens for `<s>` and `</s>` to the start and the end, but the HuggingFace tokenizer used by SentenceTransformers does.  Is the Spark implementation incorrect.

Passing the string "boring things" to each tokenizer implementation:
Spark NLP: `Array(TokenPiece(boring,boring,11775,true,0,5), TokenPiece(thing,thing,2522,true,0,4))`
SentenceTransformers: `{'input_ids': tensor([[    0, 11775,  2522,     2]]), 'attention_mask': tensor([[1, 1, 1, 1]])}`

Note the missing 0 and 2 in the Spark tokens

**Reply by maziyarpanahi** [reply]
> From running bits of the two implementations in REPLs, it looks like Spark-NLPs MPNetEmbeddings tokenizer doesn't add tokens for `<s>` and `</s>` to the start and the end, but the HuggingFace tokenizer used by SentenceTransformers does. Is the Spark implementation incorrect.
> 
> Passing the string "boring things" to each tokenizer implementation: Spark NLP: `Array(TokenPiece(boring,boring,11775,true,0,5), TokenPiece(thing,thing,2522,true,0,4))` SentenceTransformers: `{'input_ids': tensor([[ 0, 11775, 2522, 2]]), 'attention_mask': tensor([[1, 1, 1, 1]])}`
> 
> Note the missing 0 and 2 in the Spark tokens

Hi @AlexLamaison-UnlikelyAI 

The bos and eos tokens should be added internally but they are not considered as actual tokens so you won't see them in the end results.

Usually the differences between these vectors are not that big if compared within their similarity. Reasons:
- batching, you can use `.setBatchSize(1)` and test again
- sometimes there are some pre-processing going on, we'll check see if this is true for mpnet on HuggingFace (probably not)
- sometimes the strategy in truncation, padding, max length, etc. can cause this discrepancy

I am interested to see if you compare both vectors with cosine similarity how far are there from each other?

**Reply by AlexLamaison-UnlikelyAI** [reply]
You're right about the tokens.  I hadn't realised that the start and end tokens were added by the model rather than the tokenizer.

But I've checked the cosine distance and it's large.  0.62 for the "boring thing" example, [as you can see in this colab](https://colab.research.google.com/drive/1U0db6TnYsvgiICEIOz36L4sL7Z1LQKQS?usp=sharing).  This is with a batch size of 1.  Truncation/max length shouldn't apply because the text is so short.  Is there a way I can control padding with MPNetEmbeddings?

**Reply by maziyarpanahi** [reply]
Thanks @AlexLamaison-UnlikelyAI 
That is indeed a big difference! That's no longer a small discrepancy due to Python/Java float precisions. Let me ask one our teammates debug this deeper to see what has caused this issue and get back to you.

**Reply by AlexLamaison-UnlikelyAI** [reply]
Thanks @AlexLamaison-UnlikelyAI.  I'm happy to collaborate on this if it's helpful.

**Reply by AlexLamaison-UnlikelyAI** [reply]
I think I have it.  The attention mask for "boring thing" in Spark NLP ends up as `0, 1, 1, 1` where the other implementations use `1, 1, 1, 1`.  It comes down [to a line](https://github.com/JohnSnowLabs/spark-nlp/blob/8deb7da268ef7de3d7605ebed6d274d574e76a75/src/main/scala/com/johnsnowlabs/ml/ai/MPNet.scala#L220):
```val attentionMask = batch.map(sentence => sentence.map(x => if (x < this.paddingTokenId) 0L else 1L)).toArray```.

`paddingTokenId` is `1` so this inserts `0` in the mask where the start sequence token ID is 0.

I'm not sure what the correct fix is here.

**Reply by AlexLamaison-UnlikelyAI** [reply]
This screenshot is from a debugger attached to Spark-NLP just after it's called the model.  You can see the funny attention mask and the incorrect embedding.
![Screenshot from 2025-01-06 18-16-20](https://github.com/user-attachments/assets/b13d9c9e-9f6b-4d5a-9c3e-9d9854c77b94)

This screenshot is from a test case I created using the same Onnx Java code but passing in the correct mask.  When mean-pooled and normalised, this code produces the same values as calling SentenceTransformers in Python.
![Screenshot from 2025-01-06 18-15-53](https://github.com/user-attachments/assets/a76f7fbe-2fe6-46e6-b928-bc0449009686)

**Reply by AlexLamaison-UnlikelyAI** [reply, marked as answer]
This behaviour was introduced in Spark-NLP 5.5.2 [by this commit](https://github.com/JohnSnowLabs/spark-nlp/commit/4b2aa30ee76c7e47f518ee080a882e18303acd44#diff-777c70478f1c85fa42f6a697466264c0fd5ec4eb5a950e9eed488177486c77b2R220).  I've confirmed this by downgrading to 5.5.1 and then the embedding values match those produced by SentenceTransformers.

**Reply by maziyarpanahi** [reply]
Thank you @AlexLamaison-UnlikelyAI - you nailed it with your debugging! So this line here:

```
val attentionMask = batch.map(sentence => sentence.map(x => if (x < this.paddingTokenId) 0L else 1L)).toArray
```

A quick explanation, when creating attention masks for MPNet (or really any Transformer-based model), the standard rule of thumb is:
• 1 for any non-padding token.
• 0 for any padding token. 

The change was `this.paddingTokenId` instead of a hard-coded `0` which makes sense because the id for the padding can be different for different models.

Now I am not sure why is the `paddingTokenId` is also hard-coded to `1` instead of using the pad coming from the actual vocab dictionary like other features. So we should fix this and hopefully it will use the correct padding id. I will also replace `<` with `==` since we are looking for the exact id of the padding token:
<img width="169" alt="image" src="https://github.com/user-attachments/assets/cd68318e-f250-4883-ac0e-3720dbd6eaaa" />


Many thanks @AlexLamaison-UnlikelyAI for all the good work! I'll make a fix PR and will ask you to review it before the release.

**Reply by maziyarpanahi** [reply]
I can confirm, just replacing `<` with `==` will fix the issue. We are actually looking for either tokens that are exactly paddingTokenId `==` or they are not `!=`, anything else should be fine.

I will make a PR for the fix.
